src/AVLTree.h

```diff
@@ -0,0 +1,135 @@
+//
+// Created by shvmo on 29/04/2020.
+//
+
+#ifndef DATASTRUCTURESWET1_AVLTREE_H
+#define DATASTRUCTURESWET1_AVLTREE_H
+
+#include "TNode.h"
+
+
+using std::shared_ptr;
+
+template<class Key, class Info>
+class AVLTree {
+public:
+    AVLTree();
+
+    ~AVLTree();
+
+    shared_ptr<TNode<Key, Info>> getRoot() const;
+
+    shared_ptr<TNode<Key, Info>> find(const Key &key) const;
+
+    void insert(const Key &key, const Info &info);
+
+    // This is done to allow pre allocation and deal with potential allocation errors gracefully
+    void insert(shared_ptr<TNode<Key, Info>> newNode);
+
+    void remove(const Key &key);
+
+    void printTreeInfoInOrder(std::ostream &os) const;
+
+    void treeClear();
+
+    shared_ptr<TNode<Key, Info>> getNextNodeInOrder(shared_ptr<TNode<Key, Info>> node) const;
+
+    class KeyAlreadyExistsException : std::exception {
+    };
+
+    class KeyDoesNotExistException : std::exception {
+    };
+
+private:
+    shared_ptr<TNode<Key, Info>> root;
+
+    shared_ptr<TNode<Key, Info>> searchBinaryTree(const Key &key, shared_ptr<TNode<Key, Info>> treeNode) const;
+
+    /**
+  * @brief Inserts new node to binary AVL tree while keeping balance factor in check
+  *
+  * @param newNode new node to add
+  * @param treeNode Pointer of the tree root to add to
+  *
+  * @return Current root after inserting and balancing tree
+  */
+    shared_ptr<TNode<Key, Info>>
+    insertAVLTree(shared_ptr<TNode<Key, Info>> newNode, shared_ptr<TNode<Key, Info>> treeNode);
```

Comment:
can `treeNode` be renamed to `rootNode`? 
or did i misinterpret the function?

Reply:
same with `insertAVLTree()` and 
`insertTNode()` 
or 
`insertAVLSubTree()`

Reply:
(BTW: amazing function)

Reply:
fixxxed